In [1]:
import pandas as pd
from morfeusz2 import Morfeusz
morfeusz = Morfeusz() 

In [2]:
pd.set_option('display.max_colwidth', None)

## Loading data and choosing only appropiate MWE

In [3]:
poleval_df = pd.read_csv("poleval2019_task2_training_190221/index.tsv", sep="\t", header=None)
poleval_df.columns = ["phrase_id", "doc_id", "text", "lemma"]
poleval_df.head()

phrase_id  doc_id                          text  \
0      31822   99883       Toronto Dominion Centre   
1      40025   99883                       Toronto   
2     343873   99883  kompleks handlowo-kulturalny   
3      31833   99883                    Joe Fafard   
4     327365   99883                   kanadyjskim   

                          lemma  
0       Toronto Dominion Centre  
1                       Toronto  
2  kompleks handlowo-kulturalny  
3                    Joe Fafard  
4                    kanadyjski

In [4]:
print(f"Shape of the original Poleval 2019 dataset: {poleval_df.shape}")

Shape of the original Poleval 2019 dataset: (22177, 4)


In [5]:
mwe_df = poleval_df[poleval_df["text"].str.split().str.len() > 1]
mwe_df.head()

phrase_id  doc_id                                   text  \
0      31822   99883                Toronto Dominion Centre   
2     343873   99883           kompleks handlowo-kulturalny   
3      31833   99883                             Joe Fafard   
5     343872   99883              Ludwiga Mies van der Rohe   
7     343871   99883  Toronto Dominion Gallery of Inuit Art   

                                   lemma  
0                Toronto Dominion Centre  
2           kompleks handlowo-kulturalny  
3                             Joe Fafard  
5               Ludwig Mies van der Rohe  
7  Toronto Dominion Gallery of Inuit Art

In [6]:
print(f"Shape of the Poleval 2019 dataset (only multi word expressions): {mwe_df.shape}")

Shape of the Poleval 2019 dataset (only multi word expressions): (9718, 4)


**remove not polish MWE**


In [7]:
def is_polish_mwe(text):
    """
    Check if the text is a valid Polish MWE.
    Returns True if at least one word is recognized (not tagged as 'ign').
    """
    analysis = morfeusz.analyse(text)
    for _, _, interpretations in analysis:
        tag = interpretations[2]
        if tag != 'ign':
            return True
    return False

In [8]:
df_unrecognized = mwe_df[~mwe_df['text'].apply(is_polish_mwe)].reset_index(drop=True)
print(f"Number of unrecognized MWEs: {len(df_unrecognized)}")
df_unrecognized.sample(10)

Number of unrecognized MWEs: 566


phrase_id  doc_id                                text  \
16       99721  100524             Barvicha Luxury Village   
393     287881  102435                    History Carnival   
482     471925  103954                         Angry Birds   
225      72068  101751                    Creative Commons   
473     471465  103927                       Andrei Strbik   
161      67462  101384  Tiengiz Grigorjewicz Sułakwielidze   
549     490467  107326                   Nicolasem Sarkozy   
63       52562  101186                  Pearson Television   
271      73836  101886                         Two Harbors   
57       52395  101173                        Olle Hansson   

                                  lemma  
16              Barvicha Luxury Village  
393                    History Carnival  
482                         Angry Birds  
225                    Creative Commons  
473                       Andrea Strbik  
161  Tiengiz Grigorjewicz Sułakwielidze  
549                     Nicolas Sarkozy  
63                   Pearson Television  
271                         Two Harbors  
57                         Olle Hansson

In [9]:
mwe_df = mwe_df[mwe_df["text"].apply(is_polish_mwe)].reset_index(drop=True)
print(f"Shape of the Poleval 2019 dataset (only valid Polish MWEs): {mwe_df.shape}")

Shape of the Poleval 2019 dataset (only valid Polish MWEs): (9152, 4)


There was 566 fully not recognized expressions by Morfeusz.


**Remove duplicates**


In [10]:
def remove_duplicates(df):
    """
    Remove duplicates from the DataFrame based on 'text' and 'lemma'.
    Keeps the first occurrence.
    """
    return df.drop_duplicates(subset=["text", "lemma"], keep='first')

In [11]:
mwe_df = remove_duplicates(mwe_df)
print(f"Shape of the Poleval 2019 dataset (after removing duplicates): {mwe_df.shape}")

Shape of the Poleval 2019 dataset (after removing duplicates): (7453, 4)


**TO DO: podzielić na nieodmienne i mianowniki:**


In [12]:
mwe_df[mwe_df["text"] == mwe_df["lemma"]]


phrase_id  doc_id                                   text  \
0         31822   99883                Toronto Dominion Centre   
1        343873   99883           kompleks handlowo-kulturalny   
2         31833   99883                             Joe Fafard   
4        343871   99883  Toronto Dominion Gallery of Inuit Art   
5        343875  100499                        PZL.13 (PZL-13)   
...         ...     ...                                    ...   
9137     329445  107360                           Marin Leovac   
9139     329446  107360                            Roland Linz   
9142     329448  107360                              Tomas Jun   
9144     329449  107360                        Patrick Salomon   
9145     329435  107360                           Bartosz Flis   

                                      lemma  
0                   Toronto Dominion Centre  
1              kompleks handlowo-kulturalny  
2                                Joe Fafard  
4     Toronto Dominion Gallery of Inuit Art  
5                           PZL.13 (PZL-13)  
...                                     ...  
9137                           Marin Leovac  
9139                            Roland Linz  
9142                              Tomas Jun  
9144                        Patrick Salomon  
9145                           Bartosz Flis  

[3625 rows x 4 columns]

In [13]:
mwe_df = mwe_df[mwe_df["text"] != mwe_df["lemma"]]


In [14]:
print(f"Shape of the Poleval 2019 dataset (only inflected mwe): {mwe_df.shape}")


Shape of the Poleval 2019 dataset (only inflected mwe): (3828, 4)


In [15]:
mwe_df[['text', 'lemma']].sample(10)

text                               lemma
2283                             Lexusa IS                            Lexus IS
280                    igrzyskach w Nagano                   igrzyska w Nagano
3188        Sesji Sejmu Dzieci i Młodzieży             Sejm Dzieci i Młodzieży
7531                            A Johnsona                           A Johnson
1673          Anny Marii z domu Petardówny        Anna Maria z domu Petardówny
7322                     stanów kwantowych                      stany kwantowe
1746                         Willa Wrighta                         Will Wright
2100                 Osiedle XXV-lecia PRL               osiedle XXV-lecia PRL
6803  Porozumienia Międzyinstytucjonalnego  Porozumienie Międzyinstytucjonalne
1137                  departamencie Loiret                  departament Loiret

## Cleaning

In [16]:
mwe_df.isna().sum()

phrase_id    0
doc_id       0
text         0
lemma        1
dtype: int64

In [17]:
mwe_df[mwe_df["lemma"].isna()]

phrase_id  doc_id               text lemma
1334     307628  101286  San Diego Rockets   NaN

In [18]:
# I can drop it because it's not a polish MWE
mwe_df = mwe_df.dropna()

In [19]:
mwe_df.isna().sum()

phrase_id    0
doc_id       0
text         0
lemma        0
dtype: int64

In [20]:
mwe_df[mwe_df["text"] == "Dolce & Gabbana"]

phrase_id  doc_id             text                lemma
161      99725  100524  Dolce & Gabbana  Dolce &amp; Gabbana

In [21]:
mwe_df[mwe_df["text"] == "Lublin R-VIII a"]

phrase_id  doc_id             text          lemma
734      65875  101200  Lublin R-VIII a  Lublin R-VIII

In [22]:
mwe_df[mwe_df["text"] == "BBC Radio 1's Live Lounge"]

phrase_id  doc_id                       text        lemma
2264     426217  101434  BBC Radio 1's Live Lounge  BBC Radio 1

In [23]:
mwe_df[mwe_df["text"] == "Jammin' CRT.5"]

phrase_id  doc_id           text   lemma
5108     468926  102125  Jammin' CRT.5  Jammin

In [24]:
mwe_df[mwe_df["text"] == "Traxxas Rustler XL-1 '09"]


phrase_id  doc_id                      text                 lemma
6580     469441  102506  Traxxas Rustler XL-1 '09  Traxxas Rustler XL-1

In [25]:
mwe_df[mwe_df["text"] == "Żerań F S O"]


phrase_id  doc_id         text      lemma
6916     350246  102673  Żerań F S O  Żerań FSO

In [26]:
mwe_df.loc[mwe_df["text"] == "Żerań F S O", "text"] = "Żerań FSO"

In [27]:
mwe_df.loc[mwe_df["text"] == "R C S Orzeł", "text"] = "RCS Orzeł"

In [28]:
mwe_df = mwe_df[mwe_df["text"] != "S K M"]
mwe_df = mwe_df[mwe_df["text"] != "R K S"]

In [29]:
mwe_df[mwe_df["lemma"].str.contains("ZTM")]

phrase_id  doc_id   text lemma
6897     352573  102669  Z T M   ZTM

In [30]:
mwe_df[mwe_df["lemma"].str.contains("WKD")]


phrase_id  doc_id   text lemma
6988     352545  102697  W K D   WKD

In [31]:
mwe_df[mwe_df["text"] == "átlátszó ("]

phrase_id  doc_id        text     lemma
8097     481079  103921  átlátszó (  átlátszó

In [32]:
mwe_df[mwe_df["text"] == "Społecznej Rady Konsultacyjnej ds. aktualizacji „Strategii Rozwoju Bydgoszczy do 2015 roku”"]

phrase_id  doc_id  \
5224      79818  102157   

                                                                                             text  \
5224  Społecznej Rady Konsultacyjnej ds. aktualizacji „Strategii Rozwoju Bydgoszczy do 2015 roku”   

                                                                                        lemma  
5224  Społeczna Rada Konsultacyjna ds. aktualizacji„Strategii Rozwoju Bydgoszczy do 2015 roku

In [33]:
mwe_df.loc[mwe_df["text"] == 'Społecznej Rady Konsultacyjnej ds. aktualizacji „Strategii Rozwoju Bydgoszczy do 2015 roku”', "lemma"] = 'Społeczna Rada Konsultacyjna ds. aktualizacji „Strategii Rozwoju Bydgoszczy do 2015 roku"'

In [34]:
mwe_df[mwe_df["text"].str.contains("Fundacji Ośrodka")]

phrase_id  doc_id                        text                      lemma
6812      71729  102602  Fundacji Ośrodka " Karta "  Fundacja Ośrodka "Karta "

In [35]:
mwe_df.loc[mwe_df["text"] == 'Fundacji Ośrodka " Karta "', ["text", "lemma"]] = ['Fundacji Ośrodka "Karta"', 'Fundacja Ośrodka "Karta"']

In [36]:
mwe_df[mwe_df["text"].str.contains("FILLER")] # error during creating dataset

phrase_id  doc_id                        text                lemma
6973     474289  102694   Dworca FILLER Centralnego    Dworzec Centralny
6978     350524  102696  Dworzec FILLER Śródmieście  Dworzec Śródmieście

In [37]:
mwe_df.loc[mwe_df["text"] == 'Dworca FILLER Centralnego', "text"] = 'Dworca Centralnego'
mwe_df.loc[mwe_df["text"] == 'Dworzec FILLER Śródmieście', "text"] = 'Dworzec Śródmieście'

In [38]:
mwe_df[mwe_df["text"].str.contains("Herburt")]

phrase_id  doc_id                     text                     lemma
1202      50764  101263  Jan Herburt (Arłamowski  Jan Herburt (Arłamowski)

In [39]:
mwe_df[mwe_df["text"].str.contains("Naukę")]

phrase_id  doc_id       text     lemma
4075     310468  101854  Naukę 2.0  Nauka2.0

In [40]:
mwe_df.loc[mwe_df["text"] == "Naukę 2.0", "lemma"] = "Nauka 2.0"


In [41]:
mwe_df[mwe_df["text"] == "Sesji Sejmu Dzieci i Młodzieży"]

phrase_id  doc_id                            text  \
3188      63764  101638  Sesji Sejmu Dzieci i Młodzieży   

                        lemma  
3188  Sejm Dzieci i Młodzieży

In [42]:
mwe_df.loc[mwe_df["text"] == "Sesji Sejmu Dzieci i Młodzieży", "lemma"] = "Sesja Sejmu Dzieci i Młodzieży"

In [43]:
mwe_df[mwe_df["text"]=='ustawy o zmianie ustawy Kodeks postępowania karnego']

phrase_id  doc_id                                                 text  \
3279      63911  101644  ustawy o zmianie ustawy Kodeks postępowania karnego   

                                                   lemma  
3279  Ustawao zmianie ustawy Kodeks postępowania karnego

In [44]:
mwe_df.loc[mwe_df["text"]=='ustawy o zmianie ustawy Kodeks postępowania karnego', "lemma"] = 'ustawa o zmianie ustawy Kodeks postępowania karnego'


In [45]:
mwe_df[mwe_df["text"].str.contains("o ujawnieniu pracy lub służby")]


phrase_id  doc_id  \
4449      81044  101946   
4451      81046  101946   

                                                                                                                                                                               text  \
4449                    o zmianie ustawy o ujawnieniu pracy lub służby w organach bezpieczeństwa państwa lub współpracy z nimi w latach 1944–1990 osób pełniących funkcje publiczne   
4451  ustawie z dnia 11 kwietnia 1997 r. o ujawnieniu pracy lub służby w organach bezpieczeństwa państwa lub współpracy z nimi w latach 1944-1990 osób pełniących funkcje publiczne   

                                                                                                                                                                             lemma  
4449                             ustawa o ujawnieniu pracy lub służby w organach bezpieczeństwa państwa lub współpracy z nimi w latach 1944–1990 osób pełniących funkcje publiczne  
4451  Ustawa z dnia 11 kwietnia 1997 r. o ujawnieniu pracy lub służby w organach bezpieczeństwa państwa lub współpracy z nimi w latach 1944-1990 osób pełniących funkcje publiczne

In [46]:
mwe_df = mwe_df.drop(index=4449)

In [47]:
mwe_df.drop(mwe_df[mwe_df["text"].isin([
    "Dolce & Gabbana", # not a polish MWE
    "Lublin R-VIII a", # ???
    "BBC Radio 1's Live Lounge", # not a polish MWE
    "Jammin' CRT.5", # not a polish MWE
    "Traxxas Rustler XL-1 '09", # not a polish MWE
    "átlátszó (", # not a polish MWE
    "Z T M",
    "W K D"
])].index, inplace=True)

In [48]:
print(f"Shape of the Poleval 2019 dataset (after cleaning): {mwe_df.shape}")

Shape of the Poleval 2019 dataset (after cleaning): (3816, 4)


In [49]:
mwe_df.to_csv("mwe_cleaned_df.csv")

## Actual analysis

In [50]:
mwe_df = pd.read_csv("mwe_cleaned_df.csv", index_col=0)
print(f"Shape of the cleaned Poleval 2019 dataset: {mwe_df.shape}")

Shape of the cleaned Poleval 2019 dataset: (3816, 4)


In [51]:
mwe_df.head()

phrase_id  doc_id                          text                      lemma
3      343872   99883     Ludwiga Mies van der Rohe   Ludwig Mies van der Rohe
7      343879  100499       samolotu PZL.23 „Karaś”     samolot PZL.23 „Karaś”
8       31843  100499      Ministerstwa Komunikacji   Ministerstwo Komunikacji
11     343880  100499  silnikiem Pratt-Whitney Wasp  silnik Pratt-Whitney Wasp
14      31839  100499            Stanisława Praussa           Stanisław Prauss

In [52]:
mwe_df.isna().sum()

phrase_id    0
doc_id       0
text         0
lemma        0
dtype: int64

In [53]:
def is_conjunction(word):
    """
    Check if the word is a conjunction.
    Returns True if the word is a conjunction.
    """
    analysis = morfeusz.analyse(word)
    for _, _, interpretations in analysis:
        tag = interpretations[2]
        if tag.startswith('conj'):
            return True
    return False

In [54]:
def analyze_pair(words, lemmas):
    """
    Analyze a 2 word expression.
    :param words: List of 2 words in the expression.
    :param lemmas: List of 2 lemmas corresponding to the words.
    :return: number, case, relation type and head of the expression.
    """
    # possible numbers and cases for the words
    number_case = {}
    part_of_speech = {}

    # returned values
    number = None
    case = None
    relation = None
    head = None

    #print(f"Analyzing pair: {words} with lemmas: {lemmas}")

    for i, word in enumerate(words):
        number_case_i = set()
        analyses = morfeusz.analyse(word)

        for _, _, interpretations in analyses:
            tag = interpretations[2].split(":")
            if len(tag) > 2:
                number_i = tag[1]
                case_i = tag[2].split('.')

                # Morfeusz can return more than 1 possibilty of case after the dot so I make a list of it 
                for c in case_i:
                    number_case_i.add((number_i, c))

                part_of_speech[word] = tag[0]

        number_case[i] = number_case_i

    available = [i for i in number_case if number_case[i]]

    # check if it is possible to find common number and case for the words in the mwe -> ZWIĄZEK ZGODY
    common_cases = set.intersection(*number_case.values())
    if len(common_cases)>=1:
        # if there is only one common pair(number, case), then it is an agreement relation and the number and case of the expression can be assigned to the whole expression
        
        if len(common_cases) == 1:
            number_case_head_list = list(common_cases)[0]
            number = number_case_head_list[0]
            case = number_case_head_list[1]
 
        else:
            number = [x[0] for x in list(common_cases)]
            case = [x[1] for x in list(common_cases)]
            
        relation = "agreement"

        # noun -> HEAD
        for i, (word, tag) in enumerate(part_of_speech.items()):
            if tag == 'subst':
                head = lemmas[i]
                break

        # if there is no noun in the expression, then the head is the first word in the expression
            else:
                head = lemmas[0]

    elif [w_i for w_i, l_i in zip(words, lemmas) if w_i == l_i]:
        # government relation    
        for i, word in enumerate(words):
            if word == lemmas[i]:
                continue
            else: 
                if number_case[i]:
                    head = lemmas[i]
                    number_case_head_list = list(number_case[i])
                    number = [x[0] for x in number_case_head_list]
                    case = [x[1] for x in number_case_head_list]
                    relation = "government"
        
    if len(available) == 1:
        #print(f"Only one word has a valid analysis: {words[available[0]]} with lemma {lemmas[available[0]]}, words: {words}, lemmas: {lemmas}, number_case: {number_case}")
        # if there is only one word in the expression that has a valid analysis, then it is the head of the expression
        head = lemmas[available[0]]
        number_case_head_list = list(number_case[available[0]])
        number = [x[0] for x in number_case_head_list]
        case = [x[1] for x in number_case_head_list]
        
        idx_j = 1 - available[0]  # index of the second word in the pair
        if words[idx_j] == lemmas[idx_j]:
            # if the second word is the same as the lemma, then it is a government relation
            relation = "government"
        else:
            relation = "agreement"

    return number, case, relation, head


def analyze_mwe(phrase, lemma):
    """
    Analyze a multi-word expression.
    :param phrase: The multi-word expression.
    :param lemma: The lemma of the multi-word expression.
    :return: analysis results: numbers, cases, relations, heads, pairs.
    """

    phrase = phrase.replace(" - ", "-")
    phrase = phrase.replace("___", "")

    
    words = phrase.strip().split()
    lemmas = lemma.strip().split() 

    numbers = []
    cases = []
    relations = []
    heads = []
    pairs = []

    i = 0
    while i < len(words)-1:
        if is_conjunction(words[i+1]) and i + 2 < len(words):
            pair_words = [words[i], words[i+2]]
            pair_lemmas = [lemmas[i], lemmas[i+2]]
            i += 2 # skip the conjunction
        
        else:
            pair_words = words[i:i+2]
            pair_lemmas = lemmas[i:i+2]
            i += 1 

        # analyze the pair of words
        analyses = analyze_pair(pair_words, pair_lemmas)
        number, case, relation, head = analyses
        pairs.append(tuple(pair_words))
        numbers.append(number)  
        cases.append(case)
        relations.append(relation)
        heads.append(head)


    return numbers, cases, relations, heads, pairs

In [55]:
mwe_df[["numbers", "cases", "relations", "heads", "pairs"]] = mwe_df.apply(lambda row: pd.Series(analyze_mwe(row["text"], row["lemma"])), axis=1)

In [56]:
mwe_df.sample(10)

phrase_id  doc_id                                      text  \
1718     425138  101348                gminie Wieczfnia Kościelna   
2648      58422  101503                              Juliusza III   
7151     466640  102772  Instytutu Naukowo-Badawczego Kolejnictwa   
2971     467312  101573                      Don Pedrem de Mendez   
4686     349215  102030               X Liceum Ogólnokształcącego   
107       32097  100520      Papieskim Uniwersytecie Salezjańskim   
1544     422936  101315                          hrabstwie Dorset   
4542     473585  101956                        niszowych artystów   
8189     482833  103944                              Julii Marcel   
2878     356812  101534                           pracowników 50+   

                                      lemma  \
1718              gmina Wieczfnia Kościelna   
2648                            Juliusz III   
7151  Instytut Naukowo-Badawczy Kolejnictwa   
2971                    Don Pedro de Mendez   
4686              X Liceum Ogólnokształcące   
107        Papieski Uniwersytet Salezjański   
1544                        hrabstwo Dorset   
4542                        niszowi artyści   
8189                           Julia Marcel   
2878                          pracownik 50+   

                                                                                                               numbers  \
1718                                                                                                [[sg, sg, sg], sg]   
2648                                                                                                        [[sg, sg]]   
7151                                                                                                          [sg, sg]   
2971  [[sg, sg], [sg.pl, sg.pl, sg.pl, sg.pl, sg.pl, sg.pl, sg.pl], [sg.pl, sg.pl, sg.pl, sg.pl, sg.pl, sg.pl, sg.pl]]   
4686                                                                          [[sg, sg, sg, sg, sg, sg, sg], [sg, sg]]   
107                                                                                                           [sg, sg]   
1544                                                                                                            [[sg]]   
4542                                                                                                        [[pl, pl]]   
8189                                                                                                              [pl]   
2878                                                                                                        [[pl, pl]]   

                                                                                         cases  \
1718                                                                    [[voc, loc, dat], nom]   
2648                                                                              [[acc, gen]]   
7151                                                                                [gen, gen]   
2971  [[acc, nom], [inst, acc, loc, nom, gen, dat, voc], [inst, acc, loc, nom, gen, dat, voc]]   
4686                                        [[inst, acc, loc, gen, dat, voc, nom], [acc, gen]]   
107                                                                                 [loc, loc]   
1544                                                                                   [[loc]]   
4542                                                                              [[acc, gen]]   
8189                                                                                     [gen]   
2878                                                                              [[acc, gen]]   

                               relations                       heads  \
1718             [government, agreement]          [gmina, Wieczfnia]   
2648                        [government]                   [Juliusz]   
7151              [agreement, agreement]     [Instytut, Kolejnictwa]   
2971  [agreement, 

In [57]:
def extract_all_relation_data(numbers, cases, relations, heads):
    results = []

    filtered = [
        (n, c, r, h)
        for n, c, r, h in zip(numbers, cases, relations, heads)
        if n is not None and c is not None and r is not None and h is not None
    ]

    gov_data = [(n, c, r, h) for n, c, r, h in filtered if r == "government"]
    agree_data = [(n, c, r, h) for n, c, r, h in filtered if r == "agreement"]

    # GOV: only 1 pair
    if len(gov_data) == 1:
        n, c, r, h = gov_data[0]
        n_list = n if isinstance(n, list) else [n]
        c_list = c if isinstance(c, list) else [c]
        for num, case in zip(n_list, c_list):
            if num is not None and case is not None:
                results.append(("government", h, num, case))
        return results

    # GOV: more than 1 pair
    elif len(gov_data) > 1:
        all_pairs = []
        for n, c, *_ in gov_data:
            n_list = n if isinstance(n, list) else [n]
            c_list = c if isinstance(c, list) else [c]
            pairs = set(zip(n_list, c_list))
            all_pairs.append(pairs)

        common_pairs = set.intersection(*all_pairs)
        # If there are common pairs, return them
        if common_pairs:
            h = gov_data[0][3]
            for num, case in common_pairs:
                results.append(("government", h, num, case))
            return results

        # If no common pairs, return the first pair
        n = gov_data[0][0]
        c = gov_data[0][1]
        h = gov_data[0][3]
        n_list = n if isinstance(n, list) else [n]
        c_list = c if isinstance(c, list) else [c]

        for num, case in zip(n_list, c_list):
            if num is not None and case is not None:
                results.append(("government", h, num, case))
        return results


    # AGREEMENT
    if agree_data:
        all_pairs = []
        for n, c, *_ in agree_data:
            n_list = n if isinstance(n, list) else [n]
            c_list = c if isinstance(c, list) else [c]
            pairs = set(zip(n_list, c_list))
            all_pairs.append(pairs)

        common_pairs = set.intersection(*all_pairs)
        if common_pairs:
            h = agree_data[0][3]
            for num, case in common_pairs:
                results.append(("agreement", h, num, case))
            return results


        n = agree_data[0][0]
        c = agree_data[0][1]
        h = agree_data[0][3]
        n_list = n if isinstance(n, list) else [n]
        c_list = c if isinstance(c, list) else [c]

        for num, case in zip(n_list, c_list):
            if num is not None and case is not None:
                results.append(("agreement", h, num, case))
        return results

    
    return [(None, None, None, None)]  

expanded_rows = []
for idx, row in mwe_df.iterrows():
    results = extract_all_relation_data(row["numbers"], row["cases"], row["relations"], row["heads"])
    #print(results)
    for rel, head, num, case in results:
        new_row = row.copy()
        new_row["relation"] = rel
        new_row["head"] = head
        new_row["final_number"] = num
        new_row["final_case"] = case
        expanded_rows.append(new_row)

expanded_df = pd.DataFrame(expanded_rows)

In [68]:
expanded_df[160:170]

phrase_id  doc_id                                       text  \
143     309612  100523             Wambierzyckiej Królowej Rodzin   
143     309612  100523             Wambierzyckiej Królowej Rodzin   
143     309612  100523             Wambierzyckiej Królowej Rodzin   
147     305858  100523                   wojny trzydziestoletniej   
148     344031  100523                         bazyliki mniejszej   
149     344034  100523  hrabiego Franciszka Antoniego von Goetzen   
149     344034  100523  hrabiego Franciszka Antoniego von Goetzen   
150      32847  100523           Franciszka Antoniego von Goetzen   
150      32847  100523           Franciszka Antoniego von Goetzen   
151     309617  100523                                Matki Bożej   

                                    lemma  \
143           Wambierzycka Królowa Rodzin   
143           Wambierzycka Królowa Rodzin   
143           Wambierzycka Królowa Rodzin   
147               wojna trzydziestoletnia   
148                     bazylika mniejsza   
149  hrabia Franciszek Antoni von Goetzen   
149  hrabia Franciszek Antoni von Goetzen   
150         Franciszek Antoni von Goetzen   
150         Franciszek Antoni von Goetzen   
151                            Matka Boża   

                                  numbers  \
143          [[sg, sg, sg], [sg, sg, sg]]   
143          [[sg, sg, sg], [sg, sg, sg]]   
143          [[sg, sg, sg], [sg, sg, sg]]   
147                                  [sg]   
148                                  [sg]   
149  [[sg, sg], [sg, sg], [sg, sg], None]   
149  [[sg, sg], [sg, sg], [sg, sg], None]   
150            [[sg, sg], [sg, sg], None]   
150            [[sg, sg], [sg, sg], None]   
151                                  [sg]   

                                          cases  \
143          [[gen, loc, dat], [gen, loc, dat]]   
143          [[gen, loc, dat], [gen, loc, dat]]   
143          [[gen, loc, dat], [gen, loc, dat]]   
147                                       [gen]   
148                                       [gen]   
149  [[acc, gen], [acc, gen], [acc, gen], None]   
149  [[acc, gen], [acc, gen], [acc, gen], None]   
150              [[acc, gen], [acc, gen], None]   
150              [[acc, gen], [acc, gen], None]   
151                                       [gen]   

                                    relations  \
143                   [agreement, government]   
143                   [agreement, government]   
143                   [agreement, government]   
147                               [agreement]   
148                               [agreement]   
149  [agreement, agreement, government, None]   
149  [agreement, agreement, government, None]   
150             [agreement, government, None]   
150             [agreement, government, None]   
151                               [agreement]   

                                  heads  \
143                  [Królowa, Królowa]   
143                  [Królowa, Królowa]   
143                  [Królowa, Królowa]   
147                             [wojna]   
148                          [bazylika]   
149  [hrabia, Franciszek, Antoni, None]   
149  [hrabia, Franciszek, Antoni, None]   
150          [Franciszek, Antoni, None]   
150          [Franciszek, Antoni, None]   
151                             [Matka]   

                                                                                   pairs  \
143                                     [(Wambierzyckiej, Królowej), (Królowej, Rodzin)]   
143                                     [(Wambierzyckiej, Królowej), (Królowej, Rodzin)]   
143                                     [(Wambierzyckiej, Królowej), (Królowej, Rodzin)]   
147                                                        [(wojny, trzydziestoletniej)]   
148                                                              [(bazyliki, mniejszej)]   
149  [(hrabiego, Franciszka), (Franciszka, Antoniego), (Antoniego, von), (von, Goetzen)]   
149  [(hrabiego, Franciszka), (Fr

In [59]:
expanded_df.isna().sum()

phrase_id        0
doc_id           0
text             0
lemma            0
numbers          0
cases            0
relations        0
heads            0
pairs            0
relation        43
head            43
final_number    43
final_case      43
dtype: int64

In [60]:
expanded_df[expanded_df['head'].isna()]

phrase_id  doc_id                                   text  \
184      345762  100527                    archipelagu Seszeli   
407       41471  100620                          R. Brautigana   
421       40626  100626                      Sokolnikach Lesie   
536       52361  101170                         ziemie sanocką   
1152      50581  101256                     J. A. Marinsky'ego   
1156      50583  101256                       L. E. Glendenina   
1268      51007  101275                  Henrietą von Grunburg   
1532     422905  101311                             V. Vanýska   
1883     425275  101365                        paśmie Błotnego   
2201      67569  101419                         Vincenta Siewa   
2960     467268  101565                        pustynią Sahary   
2964     467290  101570                          ks. Malatesta   
3106      54580  101591                    Lewicy i Demokratów   
4159     349019  101884                          Rustlera XL-1   
4346      80983  101934                                 Dz. U.   
4732      76544  102034                            Ergo Hestii   
4889     286658  102063                 Mykoly Shpakovsky’iego   
5032     303088  102103                      SLO im. Traugutta   
5066     293694  102118                     Jamminem X1-CR RTR   
5071     293707  102119                              Hyperem 7   
5086     293734  102122                         Rustlerze XL-1   
5099     305975  102124                           Ofną Hyper 7   
5102     468914  102124                        Jamminami CRT.5   
5112     293756  102126                         Jamminów CRT.5   
5114     468930  102126                           Ofnę Hyper 7   
5116     293752  102126        Hot Bodies Lightninga 2 Pro Evo   
5119     468946  102127                         Jamminie CRT.5   
5706     473709  102255                wyścigu Paryż - Roubaix   
5722     283409  102263                            J. S. Foera   
5910     342848  102375                          J.G. Ballarda   
6435     469300  102477                     Wojciecha Kossaków   
6587     291429  102506                          Rustlera XL-1   
6628     469483  102510  bagusa Hot Bodies Lightning 2 Pro Evo   
6631     469481  102510                    bagusem Losi XXX-CR   
6633     469487  102511                         Jamminem CRT.5   
6639     292066  102512                        Jamminach CRT.5   
6643     469497  102512                         Rustlerem XL-1   
6822     311269  102602                                 Dz. U.   
6879     311267  102609                                  Dz. U   
6916     350246  102673                              Żerań FSO   
7394     338400  103677                        LO im. Staszica   
7511     338427  103699                    Kędzierzyna - Koźla   
8230     471947  103956                   Matthew B. Crawforda   

                                     lemma  \
184                     archipelag Seszele   
407                           R. Brautigan   
421                          Sokolniki Las   
536                         ziemia sanocka   
1152                        J. A. Marinsky   
1156                       L. E. Glendenin   
1268                 Henrieta von Grunburg   
1532                             V. Vanýsk   
1883                          pasmo Błotny   
2201                          Vincent Siew   
2960                       pustynia Sahara   
2964                          ks. Malatest   
3106                    Lewica i Demokraci   
4159                         Rustler  XL-1   
4346                              Dz.   U.   
4732                           Ergo Hestia   
4889                     Mykola Shpakovsky   
5032                   SLO im.   Traugutta   
5066                      Jammin X1-CR RTR   
5071                               Hyper 7   
5086                          Rustler XL-1   
5099                          Ofna Hyper 7   
5102                        Jamminam CRT.5 

In [62]:
expanded_df.shape

(6844, 13)

In [65]:
final_df = expanded_df.dropna()
print(f"Shape of the final Poleval 2019 dataset: {final_df.shape}")

Shape of the final Poleval 2019 dataset: (6801, 13)


In [67]:
final_df.to_csv("final_poleval2019_mwe_dataset.csv")

## NOtes

In [63]:
# def extract_relation_head(relations, heads):
#     for relation, head in zip(relations, heads):
#         if relation == 'government':
#             return 'governemnt', head
#         elif relation == 'agreement':
#             return 'agreement', head
#     return None, None

# mwe_df[["relation", "head"]] = mwe_df.apply(lambda row: pd.Series(extract_relation_head(row["relations"], row["heads"])), axis=1)

In [64]:
# def extract_full_info(numbers, cases, relations, heads):
#     # ignore None values
#     filtered = [
#         (n, c, r, h)
#         for n, c, r, h in zip(numbers, cases, relations, heads)
#         if n is not None and c is not None and r is not None and h is not None
#         ]

#     if not filtered:
#         return None, None, None, None

#     for n, c, r, h in filtered:
#         if r == "government":
#             if len(c)==1:
#                 return r, h, n[0], c[0]
#             else:
#                 all_numbers = [set(x) if isinstance(x, list) else {x} for x in n]

#     # Jeżeli wszystko to związki zgody
#     if all(r == "agreement" for _, _, r, _ in filtered):
#         norm_numbers = [set(x) if isinstance(x, list) else {x} for x, *_ in filtered]
#         norm_cases = [set(x) if isinstance(x, list) else {x} for _, x, *_ in filtered]

#         common_number = set.intersection(*norm_numbers)
#         common_case = set.intersection(*norm_cases)

#         head = filtered[0][3]  # pierwszy head
#         return "agreement", head, (
#             list(common_number)[0] if common_number else None
#         ), (
#             list(common_case)[0] if common_case else None
#         )


#     return None, None, None, None


# mwe_df[["relation", "head", "final_number", "final_case"]] = mwe_df.apply(
#     lambda row: pd.Series(extract_full_info(row["numbers"], row["cases"], row["relations"], row["heads"])),
#     axis=1
# )